### DISCUSSION TITLE
Github let's you overwrite on protected branch when freshly pulled repo clone

### DISCUSSION THREAD

**Discussion by ideepika** [community/community#23398] [New to GitHub]
<aside class="onebox githubissue" data-onebox-src="https://github.com/github/hub/issues/2793">
  <header class="source">

      <a href="https://github.com/github/hub/issues/2793" target="_blank" rel="noopener nofollow ugc">github.com/github/hub</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/github/hub/issues/2793" target="_blank" rel="noopener nofollow ugc">github let's you overwrite on protected branch when freshly pulled repo clone</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-07-01" data-time="18:11:28" data-timezone="UTC">06:11PM - 01 Jul 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/ideepika" target="_blank" rel="noopener nofollow ugc">
          <img alt="ideepika" src="https://user-images.githubusercontent.com/22215384/181076753-29df44a5-a95b-4236-ba62-2a08a500c3d1.jpeg" class="onebox-avatar-inline" width="20" height="20">
          ideepika
        </a>
      </div>
    </div>

    <div class="labels">
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          bug
        </span>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">this is pretty dangerous bug as it can wipe out the code if mistakenly someone f<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">orce pushes in main branch, request you to please look into the issue on priority.

I fresh cloned a repo I am a contributor to and forgot set up fork remote to push on, instead I pushed mistakenly set my remote also to point to origin/master... 
```
❯ git remote add ideepika https://github.com/ceph/ceph-build.git
❯ git push ideepika -f
Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Delta compression using up to 8 threads
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 1.05 KiB | 1.05 MiB/s, done.
Total 7 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/ceph/ceph-build.git
   9afd68cf..33e4b91f  master -&gt; master
```
This should have been rejected since the branch is protected, but it got bypassed somehow 
Since the branch was a protected branch this should have been rejected, but it still got force pushed, and when I tried to amend by repushing the original state ... I get 

```
❯ git push origin master -f
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: Cannot force-push to this protected branch
To github.com:ceph/ceph-build.git
 ! [remote rejected]   master -&gt; master (protected branch hook declined)
error: failed to push some refs to 'github.com:ceph/ceph-build.git'
```

**Command attempted:**

**What happened:**

**More info:**</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by jsoref** [comment, marked as answer]
<p>I don’t think <code>--force</code> means what you think it means.</p>
<p>Here is where your repository was:</p><aside class="onebox githubcommit" data-onebox-src="https://github.com/ceph/ceph-build/commit/9afd68cf2c530781b11d99455186b1dd59f480cb">
  <header class="source">

      <a href="https://github.com/ceph/ceph-build/commit/9afd68cf2c530781b11d99455186b1dd59f480cb" target="_blank" rel="noopener nofollow ugc">github.com/ceph/ceph-build</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M10.86 7c-.45-1.72-2-3-3.86-3-1.86 0-3.41 1.28-3.86 3H0v2h3.14c.45 1.72 2 3 3.86 3 1.86 0 3.41-1.28 3.86-3H14V7h-3.14zM7 10.2c-1.22 0-2.2-.98-2.2-2.2 0-1.22.98-2.2 2.2-2.2 1.22 0 2.2.98 2.2 2.2 0 1.22-.98 2.2-2.2 2.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/ceph/ceph-build/commit/9afd68cf2c530781b11d99455186b1dd59f480cb" target="_blank" rel="noopener nofollow ugc">Merge pull request #1861 from ceph/wip-normaldistro</a>
    </h4>

    <div class="github-info">
      <div class="date">
        committed <span class="discourse-local-date" data-format="ll" data-date="2021-06-30" data-time="21:29:06" data-timezone="UTC">09:29PM - 30 Jun 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/djgalloway" target="_blank" rel="noopener nofollow ugc">
          <img alt="djgalloway" src="https://user-images.githubusercontent.com/2119212/181076754-166b5e66-c2ec-4237-8c38-4b6c9f5ecca4.jpeg" class="onebox-avatar-inline" width="20" height="20">
          djgalloway
        </a>
      </div>

      <div class="lines" title="changed 1 files with 1 additions and 1 deletions">
        <a href="https://github.com/ceph/ceph-build/commit/9afd68cf2c530781b11d99455186b1dd59f480cb" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+1</span>
          <span class="removed">-1</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">build_utils.sh: Fix variable</p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Here is where your repository went to:</p><aside class="onebox githubcommit" data-onebox-src="https://github.com/ceph/ceph-build/commit/33e4b91f30d638b0059c4586e15afcfa090ee4a8">
  <header class="source">

      <a href="https://github.com/ceph/ceph-build/commit/33e4b91f30d638b0059c4586e15afcfa090ee4a8" target="_blank" rel="noopener nofollow ugc">github.com/ceph/ceph-build</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M10.86 7c-.45-1.72-2-3-3.86-3-1.86 0-3.41 1.28-3.86 3H0v2h3.14c.45 1.72 2 3 3.86 3 1.86 0 3.41-1.28 3.86-3H14V7h-3.14zM7 10.2c-1.22 0-2.2-.98-2.2-2.2 0-1.22.98-2.2 2.2-2.2 1.22 0 2.2.98 2.2 2.2 0 1.22-.98 2.2-2.2 2.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/ceph/ceph-build/commit/33e4b91f30d638b0059c4586e15afcfa090ee4a8" target="_blank" rel="noopener nofollow ugc">build_utils: set build flags using CMakeLists of pushed branch</a>
    </h4>

    <div class="github-info">
      <div class="date">
        committed <span class="discourse-local-date" data-format="ll" data-date="2021-07-01" data-time="17:32:34" data-timezone="UTC">05:32PM - 01 Jul 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/ideepika" target="_blank" rel="noopener nofollow ugc">
          <img alt="ideepika" src="https://user-images.githubusercontent.com/2119212/181076755-fc025a37-5ef1-4c8b-a139-8da886bcc364.jpeg" class="onebox-avatar-inline" width="20" height="20">
          ideepika
        </a>
      </div>

      <div class="lines" title="changed 2 files with 22 additions and 0 deletions">
        <a href="https://github.com/ceph/ceph-build/commit/33e4b91f30d638b0059c4586e15afcfa090ee4a8" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+22</span>
          <span class="removed">-0</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">idea is to fetch CMakeLists.txt from pushed branch on ci &gt;&gt; grep for set cmake
o<span class="show-more-container"><a href="https://github.com/ceph/ceph-build/commit/33e4b91f30d638b0059c4586e15afcfa090ee4a8" target="_blank" rel="noopener nofollow ugc" class="show-more">…</a></span><span class="excerpt hidden">n libraries &gt;&gt; export build optional for these

Signed-off-by: Deepika Upadhyay &lt;dupadhya@redhat.com&gt;</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>This is the difference:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/ceph/ceph-build/compare/9afd68cf2c530781b11d99455186b1dd59f480cb...33e4b91f30d638b0059c4586e15afcfa090ee4a8">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/ceph/ceph-build/compare/9afd68cf2c530781b11d99455186b1dd59f480cb...33e4b91f30d638b0059c4586e15afcfa090ee4a8" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181076757-6a262035-b6f1-4974-980a-702dbf856608.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/d/3db3e0d7deb01f01f07689e990839559494c4aca_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/3/d/3db3e0d7deb01f01f07689e990839559494c4aca_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/3/d/3db3e0d7deb01f01f07689e990839559494c4aca.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/d/3db3e0d7deb01f01f07689e990839559494c4aca_2_10x10.png"></div>

<h3><a href="https://github.com/ceph/ceph-build/compare/9afd68cf2c530781b11d99455186b1dd59f480cb...33e4b91f30d638b0059c4586e15afcfa090ee4a8" target="_blank" rel="noopener nofollow ugc">ceph/ceph-build</a></h3>

  <p>Helper scripts for building the official Ceph packages - ceph/ceph-build</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>If you look at the commits list for where you are now, you’ll see that the parent is where you were before.</p>
<p><code>--force</code> is used for when you’re asking git to effectively remove commits from the chain leading up to a point.</p>
<p>You aren’t doing that, and thus <code>--force</code> was ignored.</p>
<p>Say you had:</p><aside class="onebox githubcommit" data-onebox-src="https://github.com/ceph/ceph-build/commit/694e4b9ba4423ae0a3b8801248039ea5a12b8cdd">
  <header class="source">

      <a href="https://github.com/ceph/ceph-build/commit/694e4b9ba4423ae0a3b8801248039ea5a12b8cdd" target="_blank" rel="noopener nofollow ugc">github.com/ceph/ceph-build</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M10.86 7c-.45-1.72-2-3-3.86-3-1.86 0-3.41 1.28-3.86 3H0v2h3.14c.45 1.72 2 3 3.86 3 1.86 0 3.41-1.28 3.86-3H14V7h-3.14zM7 10.2c-1.22 0-2.2-.98-2.2-2.2 0-1.22.98-2.2 2.2-2.2 1.22 0 2.2.98 2.2 2.2 0 1.22-.98 2.2-2.2 2.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/ceph/ceph-build/commit/694e4b9ba4423ae0a3b8801248039ea5a12b8cdd" target="_blank" rel="noopener nofollow ugc">build_utils.sh: Fix variable</a>
    </h4>

    <div class="github-info">
      <div class="date">
        committed <span class="discourse-local-date" data-format="ll" data-date="2021-06-30" data-time="21:14:24" data-timezone="UTC">09:14PM - 30 Jun 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/djgalloway" target="_blank" rel="noopener nofollow ugc">
          <img alt="djgalloway" src="https://user-images.githubusercontent.com/2119212/181076758-55a39de3-ab8f-4ef9-8447-c2fdf97529ee.jpeg" class="onebox-avatar-inline" width="20" height="20">
          djgalloway
        </a>
      </div>

      <div class="lines" title="changed 1 files with 1 additions and 1 deletions">
        <a href="https://github.com/ceph/ceph-build/commit/694e4b9ba4423ae0a3b8801248039ea5a12b8cdd" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+1</span>
          <span class="removed">-1</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">Signed-off-by: David Galloway &lt;dgallowa@redhat.com&gt;</p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>as your <code>HEAD</code>, and you tried pushing (don’t do this, test either in a pair of local repositories, or in some unrelated test repositories in your user’s space). By default w/o <code>--force</code>, you’d get an error:</p>
<p>Here’s a long <a href="https://explainshell.com/explain?cmd=cd+%24%28mktemp+-d%29%3B+git+init+a%3B+%28cd+a%3B+touch+a%3B+git+add+a%3B+git+commit+-m+a%29%3B+git+clone+a+b%3B+%28cd+b%3B+git+checkout+-b+ignore-this%29%3B+git+clone+a+c%3B+%28cd+c%3B+git+checkout+-b+ignore-this%29%3B+%28cd+a%3B+echo+a+%3E%3E+a%3B+git+add+a%3B+git+commit+-m+A%3B+git+checkout+-b+ignore-this%3B+git+fetch+..%2Fb%3B+git+push+..%2Fb+main%29%3B+for+a+in+*%3B+do+%28cd+%24a%3B+echo+%24a%3B+git+log+main+--oneline%29%3B+done%3B+%28cd+c%3B+git+fetch+..%2Fb%3B+git+push+..%2Fb+main%3B+echo+%27well+that+did+not+work+--+try+forcing+it%21%27%3B+git+push+..%2Fb+main+--force%29%3B+for+a+in+*%3B+do+%28cd+%24a%3B+echo+%24a%3B+git+log+main+--oneline%29%3B+done" rel="noopener nofollow ugc">chain of commands</a> that you could try to follow to understand what’s happening in the basic cases:</p>
<pre><code class="lang-auto">% cd $(mktemp -d); git init a; (cd a; touch a; git add a; git commit -m a); git clone a b; (cd b; git checkout -b ignore-this); git clone a c; (cd c; git checkout -b ignore-this); (cd a; echo a &gt;&gt; a; git add a; git commit -m A; git checkout -b ignore-this; git fetch ../b; git push ../b main); for a in *; do (cd $a; echo $a; git log main --oneline); done; (cd c; git fetch ../b; git push ../b main; echo 'well that did not work -- try forcing it!'; git push ../b main --force); for a in *; do (cd $a; echo $a; git log main --oneline); done
Initialized empty Git repository in /private/var/folders/r3/n29fz25x72x191fdv6mhhr3m0000gp/T/tmp.UlNnUFSW/a/.git/
[main (root-commit) 639fd66] a
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 a
Cloning into 'b'...
done.
Switched to a new branch 'ignore-this'
Cloning into 'c'...
done.
Switched to a new branch 'ignore-this'
[main a60c634] A
 1 file changed, 1 insertion(+)
Switched to a new branch 'ignore-this'
From ../b
 * branch            HEAD       -&gt; FETCH_HEAD
Enumerating objects: 5, done.
Counting objects: 100% (5/5), done.
Writing objects: 100% (3/3), 240 bytes | 240.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
To ../b
   639fd66..a60c634  main -&gt; main
a
a60c634 (HEAD -&gt; ignore-this, main) A
639fd66 a
b
a60c634 (main) A
639fd66 (HEAD -&gt; ignore-this, origin/main, origin/HEAD) a
c
639fd66 (HEAD -&gt; ignore-this, origin/main, origin/HEAD, main) a
From ../b
 * branch            HEAD       -&gt; FETCH_HEAD
To ../b
 ! [rejected]        main -&gt; main (fetch first)
error: failed to push some refs to '../b'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
well that did not work -- try forcing it!
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To ../b
 + a60c634...639fd66 main -&gt; main (forced update)
a
a60c634 (HEAD -&gt; ignore-this, main) A
639fd66 a
b
639fd66 (HEAD -&gt; ignore-this, origin/main, origin/HEAD, main) a
c
639fd66 (HEAD -&gt; ignore-this, origin/main, origin/HEAD, main) a
</code></pre>
<p>Admittedly, the GitHub docs on allow-force-push are not end-user resilient, but I don’t have the energy to fix them.<br>
<a href="https://docs.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#allow-force-pushes" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#allow-force-pushes</a></p>
